lib/ukboot/early_init.c

```diff
@@ -0,0 +1,137 @@
+ * Keep this as a global as `argv` points to tokens in this string.
+ */
+static char cmdline_argv[CONFIG_LIBUKBOOT_CMDLINE_MAX_LEN];
+
+#if CONFIG_LIBUKLIBPARAM
+static int init_libparam(int argc, char *argv[])
+{
+	int rc;
+
+	UK_ASSERT(argv);
+
+	/*
+	 * First, we scan if we can find the stop sequence in the kernel
+	 * cmdline. If not, we assume that there are no uklibparam arguments in
+	 * the command line.
+	 * NOTE: argv[0] contains the kernel/program name that we need to
+	 *       hide from the parser.
+	 */
+	rc = uk_libparam_parse(argc - 1, &argv[1], UK_LIBPARAM_F_SCAN);
+	if (rc > 0 && rc < (argc - 1)) {
+		/* In this case, we did successfully scan for uklibparam
+		 * arguments and stop sequence is at rc < (argc - 1).
+		 */
+		/* Run a second pass for parsing */
+		rc = uk_libparam_parse(argc - 1, &argv[1], UK_LIBPARAM_F_USAGE);
+		if (unlikely(rc < 0)) /* go down on errors (including USAGE) */
+			return -EINVAL;
+
+		/* Drop uklibparam parameters from argv but keep argv[0].
+		 * We are going to replace the stop sequence with argv[0].
+		 */
+		rc += 1; /* include argv[0]; we use rc as idx to stop seq */
+		argc -= rc;
+		argv[rc] = argv[0];
+		argv = &argv[rc];
+	}
+	return 0;
+}
+#endif /* CONFIG_LIBUKLIBPARAM */
+
+static int uk_boot_early_init_parse_cmdl(struct ukplat_bootinfo *bi)
+{
+	const char *cmdl;
+	int rc __maybe_unused;
+
+	UK_ASSERT(bi);
+
+	if (!bi->cmdline_len) {
+		cmdl = CONFIG_UK_NAME;
+		cmdline_len = sizeof(CONFIG_UK_NAME);
+	} else {
+		cmdl = (const char *)bi->cmdline;
+		cmdline_len = bi->cmdline_len;
+	}
+
+	if (unlikely(cmdline_len > ARRAY_SIZE(cmdline)))
+		return -EINVAL;
+
+	/* Update bi->cmdline */
+	strncpy(cmdline, cmdl, ARRAY_SIZE(cmdline));
+	cmdline[cmdline_len] = 0;
+	bi->cmdline = (__u64)cmdline;
+
+	/* Break down cmdline to argv */
+	strncpy(cmdline_argv, cmdl, ARRAY_SIZE(cmdline));
+	cmdline_argv[cmdline_len] = 0;
+
+	boot_argc = uk_argnparse(cmdline, cmdline_len, &boot_argv[0],
+				 CONFIG_LIBUKBOOT_MAXNBARGS);
+
+#if CONFIG_LIBUKLIBPARAM
+	rc = init_libparam(boot_argc, boot_argv);
+	if (unlikely(rc))
+		return rc;
+#endif /* CONFIG_LIBUKLIBPARAM */
+
+	return 0;
+}
+
+static int uk_boot_early_init_mrd_coalesce(struct ukplat_bootinfo *bi)
+{
+	UK_ASSERT(bi);
+	ukplat_memregion_list_coalesce(&bi->mrds);
+	return 0;
+}
+
+void uk_boot_early_init(struct ukplat_bootinfo *bi)
+{
+	struct uk_boot_earlytab_entry *entry;
+	int rc;
+
+	UK_ASSERT(bi);
+
+	uk_boot_earlytab_foreach(entry) {
+		UK_ASSERT(entry);
+		rc = (*entry->init)(bi);
+		if (unlikely(rc))
+			UK_CRASH("Early init call @ %p failed with %d",
+				 entry->init, rc);
+	}
+}
+
+/* Parse cmdline parameters at the earliest priority
+ * to make them available to the rest of early init
+ */
+uk_boot_earlytab_register(uk_boot_early_init_parse_cmdl, UK_PRIO_EARLIEST);
+
+/* Reserve UK_PRIO_LATEST for post-coalesce operations */
+uk_boot_earlytab_register(uk_boot_early_init_mrd_coalesce,
```

Comment:
I guess we could add some logic, as well as to enforce a single call of coalesce. If bootinfo becomes part of `libukboot` that makes sense, otherwise if it becomes a standalone lib it seems to add too much policy so I would probably make that behavior configurable. We can keep it in mind as a feature.